### PR TITLE
[FLINK-34058][table] Support optional parameters for named parameters

### PR DESCRIPTION
**Pull request by hackergin**
## What is the purpose of the change

*Support optional named arguments*

## Brief change log

* Obtain the argument optionals information from FunctionHint or ProcedureHint.
* Modify TypeInferenceOperandChecker to support optional arguments.
* In the SqlToRel phase, convert optional arguments to Default nodes and use defaultType as RexNode type.
* During the procedure conversion process, replace optional arguments with Default nodes with default type.

## Verifying this change

  - *Added test case (call function or procedure with optional arguments) in FunctionITCase and ProcedureITCase*
  - *Add test case in TypeInferenceExtractorTest including scenarios such as obtaining optional parameters for functions and procedures, conflicts, etc.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (on)
  - The S3 file system connector: (on)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented?  Will add documentation in a separate PR.

### CODE REVIEW

**Review by xuyangzhong — COMMENTED** [review]
Thanks for driving this. I left some comments here. Further more, what about adding more tests about procedure, optional hint on parameter level, optional hint on class level, multi optional hints on multi eval functions and etc?

**Comment by xuyangzhong** [comment]
BTW, can you update the description of this pr? Currently they seem to come with the template.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java [hunk 1]
```diff
@@ -194,14 +194,17 @@ private static <T, H extends Annotation> T defaultAsNull(
                     "Argument and input hints cannot be declared in the same function hint.");
         }
 
+        Boolean[] argumentOptionals = null;
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionTemplate.java:197]: The default value is false, and we did not use defaultAsNull when obtaining the optional. I don't think there will be a situation where it exists as null.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala [hunk 2]
```diff
@@ -465,20 +467,38 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
...
+          if (rexCall.getKind == SqlKind.DEFAULT && call.getOperator
+            .isInstanceOf[BridgingSqlFunction]) => {
+        val sqlFunction = call.getOperator.asInstanceOf[BridgingSqlFunction]
+        val typeInference = sqlFunction.getTypeInference
+        val typeFactory = sqlFunction.getTypeFactory
+        if (typeInference.getTypedArguments.isPresent) {
+          val dataType = typeInference.getTypedArguments.get().get(i).getLogicalType
+          operands.set(
+            i,
+            rexCall.clone(typeFactory.createFieldTypeFromLogicalType(dataType), rexCall.operands))
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala]: Indeed, this is redundant. The type has already been corrected in the SqlToRel phase.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java [hunk 3]
```diff
@@ -286,6 +305,15 @@ private static void configureTypedArguments(
 
     private static TypeStrategy translateResultTypeStrategy(
             Map<FunctionSignatureTemplate, FunctionResultTemplate> resultMapping) {
+        if (resultMapping.size() == 1) {
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java]: Yes, after obtaining the type of default parameters in OperatorBindingCallContext, there is no need to modify it here.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala [hunk 4]
```diff
@@ -237,6 +238,9 @@ object StringCallGen {
         val currentDatabase = ctx.addReusableQueryLevelCurrentDatabase()
         generateNonNullField(returnType, currentDatabase)
 
+      case DEFAULT =>
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala:240]: Updated

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java [hunk 5]
```diff
@@ -1810,4 +1844,30 @@ public String eval(String f1, String f2) {
...
+    private static class ArgumentHintScalarFunctionNotNullTypeWithOptionals extends ScalarFunction {
+        @FunctionHint(
+                argument = {
+                    @ArgumentHint(
+                            type = @DataTypeHint("STRING NOT NULL"),
+                            name = "f1",
+                            isOptional = true),
+                    @ArgumentHint(type = @DataTypeHint("INTEGER"), name = "f2", isOptional = true)
+                })
+        public String eval(String f1, Integer f2) {
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java:2134]: After verification, there will be a NullPointerException during execution. I have added relevant detection and judgment, but this may not only be an optional issue. If the primitive is declared as a nullable type, there may be similar problems, but currently there is no similar detection. However, in order to avoid modifying the existing Function temporarily, I have not made any changes to this part. In the future, a separate JIRA can be created to track this issue.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java [hunk 6]
```diff
@@ -116,7 +127,17 @@ public Consistency getConsistency() {
...
+        if (optionalArguments.isPresent()) {
+            return optionalArguments.get().get(i);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isFixedParameters() {
+        return typeInference.getTypedArguments().isPresent();
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java]: After testing, direct modification here may have an impact on the verification of existing functions. The original default is false. If most functions are set to FixedParameter, it may have an impact on the existing function detection type verification. Currently, there are two situations, one is the implicit conversion of timestamp(3) and timestamp(6), and the other is the conversion of Pojo and RowType. Therefore, I currently change it to return true only if the optional hint is declared and there is an optional argument parameter. Otherwise, the default is false, so that the parameter verification inside the calcite framework can be skipped.


There is  failed ci link:

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=57012&view=logs&j=0c940707-2659-5648-cbe6-a1ad63045f0a&t=075c2716-8010-5565-fe08-3c4bb45824a4

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java [hunk 7]
```diff
@@ -116,7 +127,17 @@ public Consistency getConsistency() {
 
     @Override
     public boolean isOptional(int i) {
-        return false;
+        Optional<List<Boolean>> optionalArguments = typeInference.getOptionalArguments();
+        if (optionalArguments.isPresent()) {
+            return optionalArguments.get().get(i);
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/inference/TypeInferenceOperandChecker.java:138]: The argument hint is for optional parameters that have default values, so there will not be a null situation.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java [hunk 8]
```diff
@@ -46,6 +46,8 @@ public final class TypeInference {
 
     private final @Nullable List<String> namedArguments;
 
+    private final @Nullable List<Boolean> optionalArguments;
```
Comment (fsk119) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java:49]: I am worried about the current design of the Function Overload and Optional Parameter.

The current implementation suggests that a UDF can only use either function overloading or optional parameters, not both.  But in some situation, we may need both abilities.

```
class ToBytes extends ScalarFunction {

    bytes[] eval(String input, String format ="gzip") {}

    bytes[] eval(int input, String format = "gzip") {}
}
```

Actually, I think the argument optionals and names(alias) should belong to the `InputTypeStrategy`.  It's better we should open another FLIP to discuss the API design, cc @xuyangzhong @lincoln-lil

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java [hunk 9]
```diff
@@ -609,7 +611,23 @@ private static Stream<TestSpec> functionSpecs() {
                 // For function with overloaded function, argument name will be empty
                 TestSpec.forScalarFunction(
                         "Scalar function with overloaded functions and arguments hint declared.",
-                        ArgumentsHintScalarFunctionWithOverloadedFunction.class));
+                        ArgumentsHintScalarFunctionWithOverloadedFunction.class),
+                TestSpec.forScalarFunction(
+                                "Scalar function with argument type not null but optional.",
+                                ArgumentHintNotNullTypeWithOptionalsScalarFunction.class)
+                        .expectErrorMessage(
+                                "Argument at position 0 is not null but a optional argument"),
```
Comment (fsk119) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java:619]: Argument at position 0 is optional but its type deesn't accept null value.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 10]
```diff
@@ -653,6 +655,36 @@ object FlinkRexUtil {
...
+
+    extractDefaultTypes(call.getOperator, validator.getTypeFactory)
+  }
+
+  /** Extracts the default types of operands for a given SqlOperator that uses named arguments. */
+  def extractDefaultTypes(
+      sqlOperator: SqlOperator,
+      typeFactory: RelDataTypeFactory): Array[RelDataType] = {
+    sqlOperator.getOperandTypeChecker
+      .asInstanceOf[SqlOperandMetadata]
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:673]: I am a little worried about the design here. If `extractDefaultTypes` is common utils for others, it should handle all kinds of SqlOperator but the current implementation only process Operator that uses named argument. It's not a good design that users should check the implementation because others is not familiar as us.

I suggest we add a class extends SqlCallBinding and move all this into the newly added class. WDYT?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala:240]: The current implementation seems we add a built-in function named "DEFAULT" here.  Does this mean user can not create a function named "DEFAULT" anymore?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java [hunk 11]
```diff
@@ -1286,4 +1287,6 @@ public List<SqlGroupedWindowFunction> getAuxiliaryFunctions() {
                     .operandTypeChecker(OperandTypes.NILADIC)
                     .notDeterministic()
                     .build();
+
+    public static final SqlSpecialOperator DEFAULT = SqlStdOperatorTable.DEFAULT;
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:1291]: Add some description about this. Can user uses this directly?

**Review by xuyangzhong — COMMENTED** [review]
Thanks for updating. I just left some minor comments again.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala:240]: The default function is not registered, it is only used for codegen judgment.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java [hunk 12]
```diff
@@ -44,19 +45,24 @@ final class FunctionSignatureTemplate {
 
     final @Nullable String[] argumentNames;
 
+    final @Nullable Boolean[] argumentOptionals;
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java]: Currently, since optional is not included as part of inputStrategy, argumentName and argumentOptional are not involved in the calculation of equal and hashcode. Therefore, if optional participates in the calculation of hashCode and equals here, it may result in duplicate inputStrategies.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/BaseMappingExtractor.java [hunk 13]
```diff
@@ -143,6 +144,9 @@ protected Map<FunctionSignatureTemplate, FunctionResultTemplate> extractResultMa
                 // check if the method can be called
                 verifyMappingForMethod(correctMethod, collectedMappingsPerMethod, verification);
 
+                // check if we declare optional on a primitive type parameter
+                verifyOptionalOnPrimitiveParameter(correctMethod, collectedMappingsPerMethod);
+
                 // check if method strategies conflict with function strategies
                 collectedMappingsPerMethod.forEach(
                         (signature, result) -> putMapping(collectedMappings, signature, result));
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/BaseMappingExtractor.java:152]: I think we can skip this check here if the user is not using namedArgument, so I think we don't need to check it here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java [hunk 14]
```diff
@@ -44,19 +45,24 @@ final class FunctionSignatureTemplate {
...
+
     private FunctionSignatureTemplate(
             List<FunctionArgumentTemplate> argumentTemplates,
             boolean isVarArgs,
-            @Nullable String[] argumentNames) {
+            @Nullable String[] argumentNames,
+            @Nullable Boolean[] argumentOptionals) {
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java]: Removed the  `nullable` and  generating default optional for each parameter.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/BaseMappingExtractor.java [hunk 15]
```diff
@@ -323,6 +327,37 @@ private void verifyMappingForMethod(
...
+            Method method,
+            Map<FunctionSignatureTemplate, FunctionResultTemplate> collectedMappingsPerMethod) {
+
+        collectedMappingsPerMethod.forEach(
+                (signature, result) -> {
+                    Boolean[] argumentOptional = signature.argumentOptionals;
+                    if (argumentOptional != null
+                            && Arrays.stream(argumentOptional).anyMatch(Boolean::booleanValue)) {
+                        // do something check
+                        FunctionSignatureTemplate functionResultTemplate =
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/BaseMappingExtractor.java]: The main purpose here is to determine if there is an "optional" in FunctionHint while the actual method declaration is of primitive type, so it must be compared with the FunctionSignature generated by the method.

[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/BaseMappingExtractor.java]: `FunctionSignatureTemplate#of` also contains argument type. I write a typo here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/FunctionSignatureTemplate.java]: I think it's not related to InputStrategy here. `FunctionSignatureTemplate` just a POJO object that describes the method. The current implementation is counterintuitive and contradicts the way users utilize it.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java [hunk 16]
```diff
@@ -69,6 +72,10 @@ public SqlRexConvertlet get(SqlCall call) {
             return this::convertSetSemanticsWindowTableFunction;
         }
 
+        if (isContainsDefaultNode(call)) {
```
[the remark this answers is not in the document]
Comment (hackergin) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java:75]:
```java
RexNode rexCall = StandardConvertletTable.INSTANCE.convertCall(cx, call);
```

The current logic is to convert Sqlcall into RexCall first, and then correct the types of operands in RexCall. So when converting SqlCall to RexCall, callBinding still needs to convert default type. Unless we judge whether the operand is a default type before converting SqlNode into RexCall, and then correct the type, I think it should be done in  
`StandardConvertletTable.INSTANCE.convertCall`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkOperatorBinding.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The proxy implementation of {@link SqlOperatorBinding} can be used to correct the operator type
+ * when using named parameter.
+ */
+public class FlinkOperatorBinding extends SqlOperatorBinding {
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkOperatorBinding.java:44]: I think we can just extend SqlCallBinding, so we don't need to implement most of methods here.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkOperatorBinding.java [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+            int ordinal, String paramName, List<String> columnList) {
+        return sqlOperatorBinding.getColumnListParamInfo(ordinal, paramName, columnList);
+    }
+
+    @Override
+    public <T extends Object> @Nullable T getOperandLiteralValue(int ordinal, Class<T> clazz) {
+        return sqlOperatorBinding.getOperandLiteralValue(ordinal, clazz);
+    }
+
+    private List<RelDataType> getArgumentTypes() {
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkOperatorBinding.java:168]: nit: getArgumentTypes -> getNamedArgumentTypes

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 19]
```diff
@@ -6088,8 +6088,11 @@ private void translateAgg(
                 // switch out of agg mode
                 bb.agg = null;
                 // ----- FLINK MODIFICATION BEGIN -----
-                List<SqlNode> sqlNodes =
-                        new SqlCallBinding(validator(), aggregatingSelectScope, call).operands();
+                SqlCallBinding sqlCallBinding =
+                        new SqlCallBinding(validator(), aggregatingSelectScope, call);
+                List<SqlNode> sqlNodes = sqlCallBinding.operands();
+                FlinkOperatorBinding flinkOperatorBinding =
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:6094]: Can we just new a specific binding here?

File: flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 20]
```diff
@@ -6088,8 +6088,11 @@ private void translateAgg(
                 // switch out of agg mode
                 bb.agg = null;
                 // ----- FLINK MODIFICATION BEGIN -----
-                List<SqlNode> sqlNodes =
-                        new SqlCallBinding(validator(), aggregatingSelectScope, call).operands();
+                SqlCallBinding sqlCallBinding =
+                        new SqlCallBinding(validator(), aggregatingSelectScope, call);
+                List<SqlNode> sqlNodes = sqlCallBinding.operands();
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:6093]: I think we can just override the operands() here and return a special SqlNode(whose return type is not ANY)if its operator is DEFAULT.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala [hunk 21]
```diff
@@ -35,7 +35,7 @@ import org.apache.flink.table.planner.functions.bridging.BridgingSqlFunction
 import org.apache.flink.table.planner.functions.sql.FlinkSqlOperatorTable._
 import org.apache.flink.table.planner.functions.sql.SqlThrowExceptionFunction
 import org.apache.flink.table.planner.functions.utils.{ScalarSqlFunction, TableSqlFunction}
-import org.apache.flink.table.planner.plan.utils.RexLiteralUtil
+import org.apache.flink.table.planner.plan.utils.{FlinkRexUtil, RexLiteralUtil}
```
Comment (fsk119) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala:38]: Revert the change about this file.

**Review by fsk119 — APPROVED** [review]
Thanks for your update. I left some comments.
